### PR TITLE
Parallelize python validate container tests in beam_PostCommit_Py_Val…

### PR DESCRIPTION
**Pull request by y1chi**
…Cont

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by y1chi** [comment]
Run Seed Job

**Comment by y1chi** [comment]
Run seed job

File: .test-infra/jenkins/job_PostCommit_Python_ValidatesContainer_Dataflow.groovy [hunk 1]
```diff
@@ -35,11 +35,13 @@ PostcommitJobBuilder.postCommitJob('beam_PostCommit_Py_ValCont',
...
 
       // Execute shell command to test Python SDK.
-      // TODO: Parallel the script run with Jenkins DSL or Gradle.
       steps {
-        VALIDATES_CONTAINER_DATAFLOW_PYTHON_VERSIONS.each {
-          pythonVersion = it.replace('.', '')
-          shell('cd ' + commonJobProperties.checkoutDir + " && bash sdks/python/container/run_validatescontainer.sh python${pythonVersion}")
+        gradle {
+          rootBuildScriptDir(commonJobProperties.checkoutDir)
+          tasks(':sdks:python:test-suites:portable:py36:validatesContainer')
```
Comment (angoenka) [inline .test-infra/jenkins/job_PostCommit_Python_ValidatesContainer_Dataflow.groovy:41]: This list will need to be updated with every new Python version.
Shall we update the relevant doc for adding and removing python versions.

**Comment by angoenka** [comment]
cc @tvalentyn

File: .test-infra/jenkins/job_PostCommit_Python_ValidatesContainer_Dataflow.groovy (hunk 1 above)
Comment (tvalentyn) [reply .test-infra/jenkins/job_PostCommit_Python_ValidatesContainer_Dataflow.groovy:41]: I suggest we add a constant along these lines
https://github.com/apache/beam/blob/2b8938eced3b5776441246f521a38af5269f916a/sdks/python/test-suites/gradle.properties#L28

and a top-level task like so: 
https://github.com/apache/beam/blob/9f9dcd6e97b29c020f555bcaa992af56108aaa0d/sdks/python/test-suites/dataflow/build.gradle#L58

**Review by tvalentyn — CHANGES_REQUESTED** [review]
Thanks a lot for this change, @y1chi !

**Comment by tvalentyn** [comment]
Run Seed Job

File: .test-infra/jenkins/job_PostCommit_Python_ValidatesContainer_Dataflow.groovy (hunk 1 above)
Comment (y1chi) [reply .test-infra/jenkins/job_PostCommit_Python_ValidatesContainer_Dataflow.groovy]: done.

**Comment by tvalentyn** [comment]
LGTM as long as the suite (https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont_PR/162/) passes, thanks @y1chi !

**Comment by tvalentyn** [comment]
```
18:09:37     Running setup.py (path:/tmp/pip-req-build-kxmdxkxz/setup.py) egg_info for package from file:///opt/apache/beam/tars/apache-beam.tar.gz
18:09:37     Created temporary directory: /tmp/pip-pip-egg-info-cw3flwuy
18:09:37     Running command python setup.py egg_info
18:09:38     Traceback (most recent call last):
18:09:38       File "<string>", line 1, in <module>
18:09:38       File "/usr/local/lib/python3.6/tokenize.py", line 452, in open
18:09:38         buffer = _builtin_open(filename, 'rb')
18:09:38     FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pip-req-build-kxmdxkxz/setup.py'
18:09:38 WARNING: Discarding file:///opt/apache/beam/tars/apache-beam.tar.gz. Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
18:09:38 ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```
Looks like some bits of the shell script may have parallelization issues. I vaguely recall having issues with either creating virtualenv or in parallel, and or building sdk tarball sdist in parallel, which we solved by configuring dependency chains in common python tasks. 

Adding
```
    dependsOn 'installGcpTest'
    dependsOn ':sdks:python:sdist'
```
Instead of  https://github.com/y1chi/beam/blob/python_valcon/sdks/python/container/run_validatescontainer.sh#L109-L118 

might resolve it.

**Comment by y1chi** [comment]
Run Python Dataflow ValidatesContainer

**Comment by y1chi** [comment]
@tvalentyn @angoenka could you try to rerun the seed job and validate container job again?

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesContainer

**Comment by y1chi** [comment]
Run Python Dataflow ValidatesContainer

**Comment by tvalentyn** [comment]
We would need to re-run the seed job to run the suite again. The last time I ran the seed job (on #13626) and
and then the suite on this PR, I got:
```
13:38:09 Task ':sdks:python:container:buildDarwinAmd64' is not up-to-date because:
13:38:09   No history is available.
13:38:09 
13:38:09 .gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/boot.go:34:2: cannot find package "github.com/apache/beam/sdks/go/pkg/beam/artifact" in any of:
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/vendor/github.com/apache/beam/sdks/go/pkg/beam/artifact (vendor tree)
13:38:09 	/home/jenkins/.gradle/go/binary/1.12.7/go/src/github.com/apache/beam/sdks/go/pkg/beam/artifact (from $GOROOT)
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/go/pkg/beam/artifact (from $GOPATH)
13:38:09 .gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/boot.go:35:2: cannot find package "github.com/apache/beam/sdks/go/pkg/beam/model/pipeline_v1" in any of:
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/vendor/github.com/apache/beam/sdks/go/pkg/beam/model/pipeline_v1 (vendor tree)
13:38:09 	/home/jenkins/.gradle/go/binary/1.12.7/go/src/github.com/apache/beam/sdks/go/pkg/beam/model/pipeline_v1 (from $GOROOT)
13:38:09 
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/go/pkg/beam/model/pipeline_v1 (from $GOPATH)
13:38:09 > Task :sdks:python:container:buildDarwinAmd64 FAILED
13:38:09 .gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/boot.go:36:2: cannot find package "github.com/apache/beam/sdks/go/pkg/beam/provision" in any of:
13:38:09 :sdks:python:container:buildDarwinAmd64 (Thread[Execution worker for ':' Thread 4,5,main]) completed. Took 3.908 secs.
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/vendor/github.com/apache/beam/sdks/go/pkg/beam/provision (vendor tree)
13:38:09 	/home/jenkins/.gradle/go/binary/1.12.7/go/src/github.com/apache/beam/sdks/go/pkg/beam/provision (from $GOROOT)
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/go/pkg/beam/provision (from $GOPATH)
13:38:09 .gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/boot.go:37:2: cannot find package "github.com/apache/beam/sdks/go/pkg/beam/util/execx" in any of:
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/vendor/github.com/apache/beam/sdks/go/pkg/beam/util/execx (vendor tree)
13:38:09 	/home/jenkins/.gradle/go/binary/1.12.7/go/src/github.com/apache/beam/sdks/go/pkg/beam/util/execx (from $GOROOT)
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/go/pkg/beam/util/execx (from $GOPATH)
13:38:09 .gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/boot.go:38:2: cannot find package "github.com/apache/beam/sdks/go/pkg/beam/util/grpcx" in any of:
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/vendor/github.com/apache/beam/sdks/go/pkg/beam/util/grpcx (vendor tree)
13:38:09 	/home/jenkins/.gradle/go/binary/1.12.7/go/src/github.com/apache/beam/sdks/go/pkg/beam/util/grpcx (from $GOROOT)
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/go/pkg/beam/util/grpcx (from $GOPATH)
13:38:09 .gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/boot.go:39:2: cannot find package "github.com/golang/protobuf/jsonpb" in any of:
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/vendor/github.com/golang/protobuf/jsonpb (vendor tree)
13:38:09 	/home/jenkins/.gradle/go/binary/1.12.7/go/src/github.com/golang/protobuf/jsonpb (from $GOROOT)
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/golang/protobuf/jsonpb (from $GOPATH)
13:38:09 .gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/boot.go:40:2: cannot find package "github.com/golang/protobuf/proto" in any of:
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/apache/beam/sdks/python/boot/vendor/github.com/golang/protobuf/proto (vendor tree)
13:38:09 	/home/jenkins/.gradle/go/binary/1.12.7/go/src/github.com/golang/protobuf/proto (from $GOROOT)
13:38:09 	/home/jenkins/jenkins-slave/workspace/beam_PostCommit_Py_ValCont_PR/src/sdks/python/container/.gogradle/project_gopath/src/github.com/golang/protobuf/proto (from $GOPATH)
13:38:09 
13:38:09 FAILURE: Build failed with an exception.
13:38:09 
13:38:09 * What went wrong:
13:38:09 Execution failed for task ':sdks:python:container:buildDarwinAmd64'.
13:38:09 > Build failed due to return code 1 of: 
13:38:09   Command:
13:38:09    /home/jenkins/.gradle/go/binary/1.12.7/go/bin/go build -o ./build/target/launcher/darwin_amd64/boot github.com/apache/beam/sdks/python/boot
13:38:09   Env:
13:38:09 
```

**Comment by tvalentyn** [comment]
rerunning seed job just in case

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesContainer

**Comment by y1chi** [comment]
I think the goResolveDependencies issues should have been addressed, can we try the task again?

File: sdks/python/container/run_validatescontainer.sh [hunk 2]
```diff
@@ -78,23 +82,26 @@ command -v gcloud
 docker -v
 gcloud -v
 
-# Build the container
+# Verify docker image has been built.
+docker images | grep "apache/$IMAGE_NAME" | grep "$SDK_VERSION"
```
Comment (tvalentyn) [inline sdks/python/container/run_validatescontainer.sh:86]: Did you mean to do any verification logic here or this is debug output?

**Comment by tvalentyn** [comment]
Run Seed Job

**Comment by tvalentyn** [comment]
Run Python Dataflow ValidatesContainer

**Review by tvalentyn — APPROVED** [review]

**Pull request merged**